rts/motoko-rts/src/types.rs

```diff
@@ -447,29 +457,71 @@ pub type Tag = u32;
 // locations in mark-compact GC. (Reminder: objects and fields are word aligned)
 pub const TAG_OBJECT: Tag = 1;
 pub const TAG_OBJ_IND: Tag = 3;
-pub const TAG_ARRAY: Tag = 5;
-pub const TAG_BITS64: Tag = 7;
-pub const TAG_MUTBOX: Tag = 9;
-pub const TAG_CLOSURE: Tag = 11;
-pub const TAG_SOME: Tag = 13;
-pub const TAG_VARIANT: Tag = 15;
-pub const TAG_BLOB: Tag = 17;
-pub const TAG_FWD_PTR: Tag = 19; // Only used by the copying GC - not to be confused with forwarding pointer in the header used for incremental GC.
-pub const TAG_BITS32: Tag = 21;
-pub const TAG_BIGINT: Tag = 23;
-pub const TAG_CONCAT: Tag = 25;
-pub const TAG_REGION: Tag = 27;
-pub const TAG_NULL: Tag = 29;
-pub const TAG_ONE_WORD_FILLER: Tag = 31;
-pub const TAG_FREE_SPACE: Tag = 33;
+pub const TAG_ARRAY_I: Tag = 5;
+pub const TAG_ARRAY_M: Tag = 7;
+pub const TAG_ARRAY_T: Tag = 9;
+pub const TAG_BITS64_U: Tag = 11;
+pub const TAG_BITS64_S: Tag = 13;
+pub const TAG_BITS64_F: Tag = 15;
+pub const TAG_MUTBOX: Tag = 17;
+pub const TAG_CLOSURE: Tag = 19;
+pub const TAG_SOME: Tag = 21;
+pub const TAG_VARIANT: Tag = 23;
+pub const TAG_BLOB_B: Tag = 25;
+pub const TAG_BLOB_T: Tag = 27;
+pub const TAG_BLOB_P: Tag = 29;
+pub const TAG_FWD_PTR: Tag = 31; // Only used by the copying GC - not to be confused with forwarding pointer in the header used for incremental GC.
+pub const TAG_BITS32_U: Tag = 33;
+pub const TAG_BITS32_S: Tag = 35;
+pub const TAG_BITS32_F: Tag = 37;
+pub const TAG_BIGINT: Tag = 39;
+pub const TAG_CONCAT: Tag = 41;
+pub const TAG_REGION: Tag = 43;
+pub const TAG_NULL: Tag = 45;
+pub const TAG_ONE_WORD_FILLER: Tag = 47;
+pub const TAG_FREE_SPACE: Tag = 49;
 
 // Special value to visit only a range of array fields.
 // This and all values above it are reserved and mean
 // a slice of an array object (i.e. start index) for
 // purposes of `visit_pointer_fields`.
 // Invariant: the value of this (pseudo-)tag must be
 //            higher than all other tags defined above
-pub const TAG_ARRAY_SLICE_MIN: Tag = 34;
+pub const TAG_ARRAY_SLICE_MIN: Tag = 50;
+
+pub fn slice_tag(array_tag: Tag, slice_start: u32) -> Tag {
+    debug_assert!(array_tag == TAG_ARRAY_I || array_tag == TAG_ARRAY_M || array_tag == TAG_ARRAY_T);
+    debug_assert!(slice_start >= TAG_ARRAY_SLICE_MIN && slice_start < (1 << 30));
+    (((array_tag - TAG_ARRAY_I) / 2) << 30) | slice_start
```

Comment:
Here, maybe, a debug assertion could be helpful, to ensure that `array_tag - TAG_ARRAY` is even and has a distance of maximum 3 (to be encoded in two bits).
The 30 bits is also related to `MAX_ARRAY_SIZE` in `constant.rs` that it must cover the maximum size of the array, i.e. `MAX_ARRAY_SIZE < (1 << 30)`. Maybe, a local constant could be helpful (just that I double check it for the 64-bit porting, where maximum array size is a little bit larger).

Reply:
Just a thought: Maybe the odd numbering is no longer needed in future if we focus on incremental GC. In this case the division and multiplication by 2 would no longer be necessary...

Reply:
True, maybe we could do that later.
I had version where I compressed the tag just once and keep in compressed form until the final reset, but the code was less clear if slightly more efficient, so I backed out.